### PR TITLE
GHC 9.8.1 support

### PR DESCRIPTION
**Pull request by christiaanb**
This passes the testsuite on GHC 9.8.1 modulo #2559 (which fails on `master` as well)

### CODE REVIEW

File: clash-lib/src/Clash/Netlist/Id/Common.hs [hunk 1]
```diff
@@ -95,7 +95,7 @@ prettyName t = maybe t (uncurry (<>)) (maybeTuple t)
 maybeTuple :: Text -> Maybe (Text, Text)
 maybeTuple "(# #)" = Just ("Unit", "")
 maybeTuple "()" = Just ("Unit", "")
-maybeTuple t = first (\n -> "Tup" <> showt n) <$> parseTuple t
+maybeTuple t = first (\n -> "Tuple" <> showt n) <$> parseTuple t
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Netlist/Id/Common.hs:98]: Is this a cosmetic change, or something that's actually needed? Not opposed to it either way.

**Review by martijnbastiaan — COMMENTED** [review]
Cool. This is still missing CI, let me know if you want me to do anything.

File: clash-ffi/tests/Main.hs [hunk 2]
```diff
@@ -30,6 +30,9 @@
   simulator is setup correctly. These tests only ensure that data does
   not get corrupted when exchanged via the VPI.
 -}
+
+{-# LANGUAGE MonoLocalBinds #-}
```
[the remark this answers is not in the document]
Comment (christiaanb) [reply clash-ffi/tests/Main.hs:34]: Ah, right, there's a pattern match on a GADT somewhere in this file, and GHC warns that you might get "weird" type inference behavior without `MonoLocalBinds`.

File: clash-lib/src/Clash/Core/TyCon.hs [hunk 3]
```diff
@@ -125,7 +125,7 @@ isTupleTyConLike nm = tupleName (T.takeWhileEnd (/= '.') (nameOcc nm))
       | Just ('(', nm1) <- T.uncons nm0
       , Just (nm2, ')') <- T.unsnoc nm1
       = T.all (== ',') nm2
-    tupleName _ = False
+    tupleName _ = T.pack "GHC.Tuple.Prim.Tuple" `T.isPrefixOf` (nameOcc nm)
```
[the remark this answers is not in the document]
Comment (christiaanb) [reply clash-lib/src/Clash/Core/TyCon.hs:128]: I don't think that works:
```
ghci> show (''())
"GHC.Tuple.Prim.Unit"
ghci> show (''(,))
"GHC.Tuple.Prim.Tuple2"
```

File: clash-lib/src/Clash/Netlist/Id/Common.hs (hunk 1 above)
Comment (christiaanb) [reply clash-lib/src/Clash/Netlist/Id/Common.hs:98]: So in GHC 9.8.1, we no longer have as the original tuple type constructor names `(,)`, `(,,)`,  `(,,,)`, etc. but instead we have `Tuple2`, `Tuple3, `Tuple4` etc.

The above code translated `(,)` to `Tup2`.

Now, in order for all of our code to produce the same output regardless whether were in an environment with GHC 9.8 or a version older than that, we have Clash translate `(,)` to `Tuple2` when we're on an older GHC versions.

Hope that makes sense.

**Comment by christiaanb** [comment]
@martijnbastiaan With regards to CI, I guess we could add https://downloads.haskell.org/ghc/9.8.1-alpha2/ and only run it during nightlies? Once 9.8.1 is released properly we can then run it against every PR.

File: clash-lib/src/Clash/Core/TyCon.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/TyCon.hs:128]: Ah makes sense, thanks!

File: clash-lib/src/Clash/Netlist/Id/Common.hs (hunk 1 above)
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Netlist/Id/Common.hs:98]: Yes, makes sense, thanks

**Comment by martijnbastiaan** [comment]
Gotcha. I'm fine with waiting around for 9.8.1 proper and merging this regardless.

**Review by martijnbastiaan — APPROVED** [review]

**Pull request merged**